### PR TITLE
Make TabBar Accessible

### LINKED ISSUE #345: Can't tab into TabBar or Paginator
Keyboard interactions don't work as expected with these components.

### PR DESCRIPTION
**Pull request by chawes13**
https://www.w3.org/TR/wai-aria-practices-1.1/#tabpanel

Resolves #345

### CODE REVIEW

**Comment by dpikt** [comment]
@chawes13 I don't think the `defaultValue` prop is necessary (the default to first tab makes sense), if another default is desired it can accomplished by using `value`.

It might be an interesting enhancement to make this input "controllable": https://medium.com/myheritage-engineering/how-controllable-react-components-maximize-reusability-86e3d233fa8e. If we want to keep it controlled, though, we should probably make `value` required.

**Comment by dpikt** [comment]
That being said, I'm pretty sure that React docs article I linked in #388  was about NOT doing precisely what this article suggests... 🤔

File: src/controls/tab-bar/focus.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
+const PERTINENT_KEY_CODES = {
+  DOWN: '40',
+  UP: '38',
+  HOME: '36',
+  END: '35',
+  LEFT: '37',
+  RIGHT: '39',
+}
+
+function getClosestTabList (el) {
```
Comment (dpikt) [inline src/controls/tab-bar/focus.js]: Can you add some brief comments about what the functions in this file do?

**Comment by chawes13** [comment]
> That being said, I'm pretty sure that React docs article I linked in #388 was about NOT doing precisely what this article suggests... 🤔

@dpikt Can we meet on this? It's a little over my head at the minute

**Comment by chawes13** [comment]
@dpikt this is ready for review! I bumped it as a `minor` version given the newly added behavior to default to the first tab. The other fixes would probably fit under `patch` - it's a grey area.

One thing that seemed really weird to me is that `options` is an optional argument. Should we modify this in #311 ?

File: src/controls/tab-bar/focus.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+function getSafeIndex (length, newIndex) {
+  const maxIndex = length - 1
+  
+  if (newIndex < 0) return getSafeIndex(length, length + newIndex)
+  if (newIndex > maxIndex) return getSafeIndex(length, length - newIndex)
+  
+  return newIndex
+}
+
+function createFocusListener (alignment='horizontal') {
```
Comment (dpikt) [inline src/controls/tab-bar/focus.js]: @chawes13 can you comment this one too? 😇

Comment (chawes13) [reply src/controls/tab-bar/focus.js]: Doneso!

**Comment by chawes13** [comment]
@dpikt LMK if you have any more feedback

File: src/controls/tab-bar/focus.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+function focusExtremeControl (control, top=true) {
+  const tabList = getClosestTabList(control)
+  const tabs = Array.from(tabList.querySelectorAll('[role="tab"]'))
+  
+  top ? tabs[0].focus() : tabs[tabs.length - 1].focus()
+}
+
+// Recursively calculates a "safe" index by wrapping around when the array bounds are exceeded
+// e.g., getSafeIndex([1, 2, 3].length, 3) -> 0
+function getSafeIndex (length, newIndex) {
```
Comment (dpikt) [inline src/controls/tab-bar/focus.js]: Can use the modulus operator for this! `safeIndex = index % array.length`

File: src/controls/tab-bar/focus.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  UP: '38',
+  HOME: '36',
+  END: '35',
+  LEFT: '37',
+  RIGHT: '39',
+}
+
+// Recursively searches for the closest parent tab list
+function getClosestTabList (el) {
+  return el && (el.matches('[role="tablist"]') ? el : getClosestTabList(el.parentElement))
```
Comment (dpikt) [inline src/controls/tab-bar/focus.js]: This double conditional is a little hard to read.

File: src/controls/tab-bar/focus.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    
+    const key = (e.which || e.keyCode || '').toString()
+    switch (key) {
+      case PERTINENT_KEY_CODES.DOWN:
+      case PERTINENT_KEY_CODES.UP: {
+        if (!vertical) break
+        
+        e.preventDefault()
+        const direction = (key === PERTINENT_KEY_CODES.DOWN) ? 1 : -1
+        focusNextControl(e.target, direction)
```
Comment (dpikt) [inline src/controls/tab-bar/focus.js]: These arguments are very flag-like, I'd suggest creating separate functions and DRYing within the functions themselves:
```js
focusNextControl()
focusPreviousControl()
focusBottomControl()
focusTopControl()
```

File: src/controls/tab-bar/focus.js (hunk 3 above)
Comment (chawes13) [reply src/controls/tab-bar/focus.js]: Does that handle the case where the index is negative? e.g., `fn([0, 1, 2], -1) -> 2`

Comment (dpikt) [reply src/controls/tab-bar/focus.js]: Looks like you need a floored modulo - so a helper is still necessary ☹️ https://stackoverflow.com/a/43827557

Comment (chawes13) [reply src/controls/tab-bar/focus.js]: 🤯

Comment (dpikt) [reply src/controls/tab-bar/focus.js]: By the way, ruby (and lots of other languages) would do that by default. Also I found this neat function too (it doesn't wrap though): https://lodash.com/docs/4.17.15#nth

File: src/controls/tab-bar/focus.js (hunk 5 above)
Comment (chawes13) [reply src/controls/tab-bar/focus.js]: Thanks for the feedback! I've incorporated this into `focus.js` and it has significantly improved the quality of the code imo. If you have time, please take a look and let me know if you would have done anything differently!

**Pull request merged**